File: Solutions/33046/c++/main.cpp
#include <iostream>
#include <vector>
#include <queue>

class MaxConnectedComponentSizeFinder {
private:
    int numberOfNodes;
    std::vector<std::vector<int>> adjacencyList;

public:
    MaxConnectedComponentSizeFinder(int number_of_nodes) : numberOfNodes(number_of_nodes),
                                                           adjacencyList(number_of_nodes + 1) {}

    void addEdge(int nodeA, int nodeB) {
        adjacencyList[nodeA].push_back(nodeB);
        adjacencyList[nodeB].push_back(nodeA);
    }

    int findMaxComponentSize() {
        std::queue<int> queue;
        std::vector<int> visited(numberOfNodes + 1, 0);
        int maxSize = -1;

        queue.push(1);

        while (!queue.empty()) {
            int currentNode = queue.front();
            queue.pop();

            if (visited[currentNode] != 0) {
                continue;
            }

            visited[currentNode] = 1;
            int currentComponentSize = 0;

            for (int neighbor : adjacencyList[currentNode]) {
                currentComponentSize++;
                queue.push(neighbor);
            }

            maxSize = std::max(maxSize, currentComponentSize);
        }

        return maxSize;
    }
};

int main() {
    int numberOfNodes;
    std::cin >> numberOfNodes;

    MaxConnectedComponentSizeFinder maxConnectedComponentFinder(numberOfNodes);

    for (int i = 0; i < numberOfNodes - 1; i++) {
        int nodeA, nodeB;
        std::cin >> nodeA >> nodeB;
        maxConnectedComponentFinder.addEdge(nodeA, nodeB);
    }

    int result = maxConnectedComponentFinder.findMaxComponentSize();
    std::cout << result << std::endl;

    return 0;
}

File: Solutions/33046/python/main.py
from collections import deque


class MaxConnectedComponentSizeFinder:
    def __init__(self, number_of_nodes):
        self.number_of_nodes = number_of_nodes
        self.adjacency_list = [[] for _ in range(number_of_nodes + 1)]

    def add_edge(self, node_a, node_b):
        self.adjacency_list[node_a].append(node_b)
        self.adjacency_list[node_b].append(node_a)

    def find_max_component_size(self):
        queue = deque([1])
        visited = [0] * (self.number_of_nodes + 1)
        max_size = -1

        while queue:
            current_node = queue.popleft()
            if visited[current_node] != 0:
                continue

            visited[current_node] = 1
            current_component_size = 0

            for neighbor in self.adjacency_list[current_node]:
                current_component_size += 1
                queue.append(neighbor)

            max_size = max(max_size, current_component_size)

        return max_size


if __name__ == "__main__":
    number_of_nodes = int(input())
    max_connected_component_finder = MaxConnectedComponentSizeFinder(number_of_nodes)

    for _ in range(number_of_nodes - 1):
        node_a, node_b = map(int, input().split())
        max_connected_component_finder.add_edge(node_a, node_b)

    result = max_connected_component_finder.find_max_component_size()
    print(result)
